energy-integration/governance-v2/src/lib.rs

```diff
@@ -25,34 +27,76 @@ pub trait GovernanceV2:
         self.require_caller_not_self();
         self.require_valid_proposal_id(proposal_id);
 
-        let depositor_mapper = self.payments_depositor(proposal_id);
-        require!(depositor_mapper.is_empty(), "Payments already deposited");
-
-        let required_payments = self.required_payments_for_proposal(proposal_id).get();
         require!(
-            !required_payments.is_empty(),
-            "This proposal requires no payments"
+            !self.proposal_reached_min_fees(proposal_id),
+            MIN_FEES_REACHED
         );
 
-        let actual_payments = self.call_value().all_esdt_transfers();
+        let additional_fee = self.call_value().single_esdt();
         require!(
-            actual_payments == required_payments,
-            "Invalid payments, must match the required payments"
+            self.governance_token_id()
+                .get()
+                .eq(&additional_fee.token_identifier),
+            WRONG_TOKEN_ID
         );
 
         let caller = self.blockchain().get_caller();
-        depositor_mapper.set(&caller);
+        let mut proposal = self.proposals().get(proposal_id);
+        proposal.fees.entries.push(FeeEntry {
+            depositor_addr: caller.clone(),
+            tokens: additional_fee.clone(),
+        });
+        proposal.fees.total_amount += additional_fee.amount.clone();
+
+        self.proposals().set(proposal_id, &proposal);
+        self.user_deposit_event(&caller, proposal_id, &additional_fee);
+    }
+
+    /// Used to claim deposited tokens to gather threshold min_fee.
+    #[payable("*")]
+    #[endpoint(claimDepositedTokens)]
+    fn claim_deposited_tokens(&self, proposal_id: ProposalId) {
+        self.require_caller_not_self();
+        self.require_valid_proposal_id(proposal_id);
+
+        require!(
+            !self.proposal_reached_min_fees(proposal_id),
+            MIN_FEES_REACHED
+        );
+        let caller = self.blockchain().get_caller();
+        let mut proposal = self.proposals().get(proposal_id);
+
+        for fee_entry in proposal.fees.entries.iter() {
+            if caller == fee_entry.depositor_addr {
+                let payment = fee_entry.tokens;
+                self.send().direct_esdt(
+                    &fee_entry.depositor_addr,
+                    &payment.token_identifier,
+                    payment.token_nonce,
+                    &payment.amount,
+                );
 
-        self.user_deposit_event(&caller, proposal_id, &actual_payments);
+                let index = proposal
+                    .fees
+                    .entries
+                    .iter()
+                    .position(|fee_entry| fee_entry.depositor_addr == caller)
+                    .unwrap();
+                proposal.fees.entries.remove(index);
+
+                self.proposals().set(proposal_id, &proposal);
+                self.user_claim_deposited_tokens_event(&caller, proposal_id, &payment);
+                return;
```

Comment:
There are a couple problems with this:
- first, this is a very inefficient way of doing this. First, we go over the vec one by one, then we perform an additional search for the item again. 
- second, we assume each user has only one position, which is not enforced

Rewrite this using a `while` loop, something like:
let mut i = 0;
while i < vec.len() {
     if saved_user == caller { 
          vec.remove(i);
          // save payment to another vec, to send at the end 
      }
     else { i+= 1; } 
}